driver-core/src/main/com/mongodb/internal/async/AsyncRunnable.java

```diff
@@ -0,0 +1,165 @@
+     * @param callback the callback
+     */
+    default void completeAlways(final Runnable runnable, final SingleResultCallback<Void> callback) {
+        this.complete((r, e) -> {
+            try {
+                runnable.run();
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(r, e);
+        });
+    }
+
+    /**
+     * @param runnable The async runnable to run after this one
+     * @return the composition of this and the runnable
+     */
+    default AsyncRunnable run(final AsyncRunnable runnable) {
+        return (c) -> {
+            this.complete((r, e) -> {
+                if (e != null) {
+                    c.onResult(null, e);
+                    return;
+                }
+                try {
+                    runnable.complete(c);
+                } catch (Throwable t) {
+                    c.onResult(null, t);
+                }
+            });
+        };
+    }
+
+    /**
+     * @param supplier The supplier to supply using after this runnable.
+     * @return the composition of this runnable and the supplier
+     * @param <T> The return type of the supplier
+     */
+    default <T> AsyncSupplier<T> supply(final AsyncSupplier<T> supplier) {
+        return (c) -> {
+            this.complete((r, e) -> {
+                if (e != null) {
+                    c.onResult(null, e);
+                    return;
+                }
+                try {
+                    supplier.complete(c);
+                } catch (Throwable t) {
+                    c.onResult(null, t);
+                }
+            });
+        };
+    }
+
+    /**
+     * @param errorCheck A check, comparable to a catch-if/otherwise-rethrow
+     * @param runnable   The branch to execute if the error matches
+     * @return The composition of this, and the conditional branch
+     */
+    default AsyncRunnable onErrorIf(
+            final Function<Throwable, Boolean> errorCheck,
+            final AsyncRunnable runnable) {
+        return (callback) -> this.complete((r, e) -> {
+            if (e == null) {
+                callback.onResult(r, null);
+                return;
+            }
+            try {
+                Boolean check = errorCheck.apply(e);
+                if (check) {
+                    runnable.complete(callback);
+                    return;
+                }
+            } catch (Throwable t) {
+                callback.onResult(null, t);
+                return;
+            }
+            callback.onResult(r, e);
+        });
+    }
+
+    /**
+     * @see RetryingAsyncCallbackSupplier
+     * @param shouldRetry condition under which to retry
+     * @param runnable the runnable to loop
+     * @return the composition of this, and the looping branch
+     */
+    default AsyncRunnable runRetryingWhen(
+            final Predicate<Throwable> shouldRetry,
+            final AsyncRunnable runnable) {
```

Comment:
Let's switch these two parameters places to make the async code look more similar to the sync counterpart. Currently, this affects the `OidcAuthenticator.authLockAsync` method.